### ISSUE TITLE
Simulator startup extremely slow (Python CPU utilization 100%)

### ISSUE THREAD

**Issue by Suizi11** [Azure/iotedgehubdev#287]
For about 6 weeks the simulator startup process is very, very slow. First, the tool tries to login at the container registry, stops the edgeHubDev and removes all the containers from previous runs. After the network check (or whatever this is intended for) the process hangs for about 1 minute. During this time I've one "Python" process in the activity monitor (macOS) which uses 100% of the CPU the whole time. After waiting a minute, everything starts as expected

```
Login Succeeded
Stopping edgeHubDev ... done
Removing <container>            ... done
Removing <container>            ... done
Removing <container>            ... done
Removing edgeHubDev          ... done
Network azure-iot-edge-dev is external, skipping
```
--> here anything weird is happening with the process, then it proceeds as always (before)
```
Pulling edgeHubDev ... done
...
```

- $ python3 --version
Python 3.8.5
- $ pip3 --version
pip 20.2.3 from /usr/local/lib/python3.8/site-packages/pip (python 3.8)
- $ iotedgehubdev --version
iotedgehubdev, version 0.14.1
- $ docker --version
Docker version 19.03.12, build 48a6621

**Comment by marianan** [comment]
Thank you filing the issue and sharing the logs - I don't see any issues with the logs you shared. Could you try to install some of the previous releases to see if the performance is consistently better - this will help narrow down any possible regressions.

**Comment by Suizi11** [comment]
Unfortunately not, python process still at ~100% and startup takes about 3 minutes.

<img width="1134" alt="image" src="https://user-images.githubusercontent.com/11649054/95581319-1401ad80-0a39-11eb-80a3-48bfc3c8604b.png">

I've installed version 0.11.1, because this one isn't available https://pypi.org/project/iotedgehubdev/0.13.0/

`pip3 install iotedgehubdev==0.13.0
ERROR: Could not find a version that satisfies the requirement iotedgehubdev==0.13.0 (from versions: 0.1.0, 0.1.1, 0.2.0, 0.3.0, 0.4.0, 0.5.0, 0.6.0, 0.7.0, 0.8.0, 0.9.0, 0.10.0, 0.11.0, 0.11.1, 0.14.0, 0.14.1)
ERROR: No matching distribution found for iotedgehubdev==0.13.0
`

**Comment by Suizi11** [comment]
Hey @marianan - I think I've found some important additional information for you.

I can only reproduce this behavior if a "Binds" array is provided in the deployment manifest, e.g. my create options look like as following:
```
{
    "Binds":[
        "<secret-host-source-path>:/mnt/host",
        "/Users/<secret-path>/modules/WebServer/src:/app/src"
    ]
}
```

If I remove the binds object everything works fine and fast.

**Comment by konichi3** [comment]
The issue is in our backlog now. Will circle back when we have an update

**Comment by glneo** [comment]
Looks like it is getting jammed up running the regex here:
https://github.com/Azure/iotedgehubdev/blob/f5fbdc3a390a9b010e1e2320ff828d0945fdde3a/iotedgehubdev/compose_parser.py#L190
As I only have the simple path name cases, my temp workaround is to replace that line with:

`match = None`

causing it to always fallback to the simple token based parsing below.

The proper fix would be to find a better regex that doesn't take 3 minutes to process for each bind.

**Comment by marianan** [comment]
Thank you @glneo the bug is in our backlog.

**Comment by Suizi11** [comment]
@marianan - do you've a rough estimation when this issue will be solved or when somebody starts working on it respectively?

Thanks a lot.

**Comment by cindydeng1998** [comment]
Hi @Suizi11, thank you for sharing your findings with us. We have done some investigation on this issue and have decided to leave the regex parsing @glneo mentioned above as is. 

This is because we need to have a thorough way of parsing binds for everyone that uses our tool, even though it is the source of a longer runtime and high CPU usage. The regex pattern here ([iotedgehubdev regex](https://github.com/Azure/iotedgehubdev/blob/73de8a10fd9c66d3c71105e5b528a43547c696ba/iotedgehubdev/constants.py#L43-L53)) was taken directly from [how docker processes binds](https://github.com/docker/docker-ce/blob/f9756bfb29877236a83979170ef2c0aa35eb57c6/components/engine/volume/mounts/windows_parser.go#L19-L76). 

Thank you very much.

**Comment by marianan** [comment]
@glneo thank you for your PR. We will review promptly.

**Comment by marianan** [comment]
@Suizi11 a new version (0.14.13) of the iotedgehubdev simulator has been released with a fix for the createOptions path array binding. Please try it out and let us know if your original issue still persists.

**Comment by ljani** [comment]
`0.14.13` fixed the issue for me. However, I had to install that `regex` dependency manually, not sure why `pip` didn't install it for me. Though I forced the `iotedgehubdev@0.14.13` install with `iotedgedev@3.3.3`.

**Comment by marianan** [comment]
Thanks @ljani i missed adding regex as a dependency in the setup.py in my earlier change. An update is coming early next week which will include also this fix.

**Comment by marianan** [comment]
@ljani the missing regex dependency is now resolved in version 0.14.4. Let us know if you run into other issues.

If you have a few mins we would love to hear your feedback on your developer experience in general at http://aka.ms/edgedevsurvey.

**Comment by marianan** [comment]
Closing this issue, please let us know if any of the symptoms persist.
